dashboard/src/components/App.js

```diff
@@ -0,0 +1,347 @@
+        }
+      });
+
+      if (connectionStatus == true) {
+        M.toast({
+          html: device.name + ' connected'
+        });
+        return oldConnectedDeviceList.concat(device);
+      } else {
+        return oldConnectedDeviceList;
+      }
+    });
+  }
+
+  function onDataWithNewIdReceived(id) {
+    // first recording of data with that ID
+    console.log('Received first recording of data with ID: ' + id);
+
+    var ids = dataRecordingContainer.getIds().sort();
+    var optionText = DataRecordingContainer.getReadableId(id);
+
+    updateDataList([]);
+
+    // append available IDs as options
+    ids.forEach(id => {
+      var optionValue = id;
+      var optionText = DataRecordingContainer.getReadableId(id);
+
+      updateDataList((oldDataList) => {
+        return oldDataList.concat({
+          id: optionValue,
+          optionText: optionText
+        });
+      });
+    });
+  }
+
+
+  // HANDLE SELECT-CHANGE
+  function handleDeviceChange(event) {
+
+    setupSocket();
+
+    var selectedDeviceId = event.target.value;
+    selectedDevice = connectedDevices.filter(connectedDevice => connectedDevice.id == selectedDeviceId)[0];
+    console.log('Selected device changed: ' + JSON.stringify(selectedDevice));
+    updateDataRecordingContainer(new DataRecordingContainer());
+  }
+
+  function handleDataChange(event) {
+
+    updateSelectedDataId(event.target.value);
+    console.log('Selected data ID changed: ' + event.target.value);
+    updateRecreateChartPlot(true);
+  }
+
+
+  return (
+    <div className="section">
+    <Logo />
+      <div className="row">
+        <Visualization dataRecordingContainer={dataRecordingContainer} selectedDataId={selectedDataId} />
+        <Row>
+          <Device deviceList={connectedDevices} dataList={dataList} handleDeviceChange={handleDeviceChange} handleDataChange={handleDataChange} />
+          <Info />
+        </Row>
+        {/* <ModalView /> */}
+      </div>
+    </div>
+  );
+}
+
+
+class DataRecordingContainer {
+
+  constructor() {
+    this.dataRecordings = {};
+  }
+
+  get dataRecordings() {
+    return this._dataRecordings;
+  }
+
+  set dataRecordings(dataRecordings) {
+    this._dataRecordings = dataRecordings;
+  }
+
+  getIds() {
+    return Object.keys(this.dataRecordings);
+  }
+
+  getData(id) {
+    if (!(id in this.dataRecordings)) {
+      this.dataRecordings[id] = [];
+    }
+    return this.dataRecordings[id];
+  }
+
+  setData(id, data) {
+    this.dataRecordings[id] = data;
+  }
+
+  addData(id, data) {
+    this.setData(id, this.getData(id).concat(data));
+  }
+
+  addDataRecording(dataRecording) {
+    this.addData(dataRecording.dataId, dataRecording.dataList);
+  }
+
+  getDataValuesInDimension(id, dimension) {
+    var values = [];
+    var maximumAggregationTimestamp = Date.now() - MINIMUM_DATA_AGE - timestampOffset;
+    this.getData(id)
+      .filter(data => data.aggregationTimestamp < maximumAggregationTimestamp)
+      .forEach(data => {
```

Comment:
instead of pushing onto new array you could use map and direct return.

e.g.:

```
return this.getData(id)
      .filter(data => data.aggregationTimestamp < maximumAggregationTimestamp)
      .map(data => {
        let value;
        if (data.hasOwnProperty('value')) {
          value = data.value;
        } else {
          const firstValue = data.values[0];
          if (firstValue instanceof Array) {
            value = firstValue[dimension];
          } else {
            value = data.values[dimension];
          }
        }
        return value;
      });
```